### DISCUSSION TITLE
Recommended way to get a percentage?

### DISCUSSION THREAD

**Discussion by decafboba** [PaulRBerg/prb-math#73] [Q&A] [answered]
For uint256 specifically, let's say I have some uint256 with a value (ie, ` 131760045705399555255526021`) and I want to take 25% of it in the most precise way possible, what would be the best way to do this via this library?

If this is a common use-case, on small request might be to have a `perc` function.

Thanks!

**Comment by PaulRBerg** [comment, marked as answer]
Hey @kenvogg, there are [examples](https://github.com/paulrberg/prb-math/tree/e2b59d8304e7c69accbabeb7c28fe76eba3de32e#prbmathud60x18sol) in the README.

In your case, you would write:

```solidity
function takePercentage(uint256 x) external pure returns (uint256 result) {
  const percentage = .25e18; // 0.25 means 25%
  result = x.mul(percentage);
}
```

To be honest I don't see the value of a `perc` function. It'd be just syntactic sugar for `mul`.

**Reply by decafboba** [reply]
Thanks @paulrberg!

Also wondering what the method would be for TypeScript BN evm-bn. I tried a bunch of shenanigans to try to get `.25e18` to work there but no luck.  Not:
```
toBn("25e18")
```
nor
```
div(toBn("1", 18), toBn("1000000000000000000", 18));
```
Almost seems like values below "1" are impossible for TypeScript BN, is that correct?

**Reply by PaulRBerg** [reply]
This is a question for the [evm-bn discussions channel](https://github.com/paulrberg/evm-bn/discussions). But, for the sake of helping you quickly, here's the answer:

```typescript
const foo = toBn("25"); // first option
const bar = toBn("25", 18); // first option
```

By default, the `toBn` function applies 18 decimals to the input. Refer to the README of the [evm-bn](https://github.com/paulrberg/evm-bn) repo for more details.

> Almost seems like values below "1" are impossible for TypeScript BN, is that correct?

No. `toBn("1")` would yield 1e18, which is 1.

**Reply by decafboba** [reply]
> Hey @kenvogg, there are [examples](https://github.com/paulrberg/prb-math/tree/e2b59d8304e7c69accbabeb7c28fe76eba3de32e#prbmathud60x18sol) in the README.
> 
> In your case, you would write:
> 
> ```solidity
> function takePercentage(uint256 x) external pure returns (uint256 result) {
>   const percentage = .25e18; // 0.25 means 25%
>   result = x.mul(percentage);
> }
> ```
> 
> To be honest I don't see the value of a `perc` function. It'd be just syntactic sugar for `mul`.

If the percentage was not fixed (so cannot assume 25% in all cases), this way wouldn't work because floating points can only be const, right?  What would be the recommended solution in that case?

**Reply by PaulRBerg** [reply]
I don't understand what you mean, PRBMath doesn't work with floating points, just fixed points.

Could you please restate your question?

**Reply by decafboba** [reply]
What I mean is, instead of the example takePercentage with fixed 25%, it would be:
```
function takePercentage(uint256 x, uint256 percentage) external pure returns (uint256 result) {
  result = x.mul(percentage);
}
```
And because a const can't be used, this wouldn't work as far as I know. So I know believe the next best way to get a percentage is:
```
function takePercentage(uint256 x, uint256 percentage) external pure returns (uint256 result) {
  result = x.mul(percentage).div(100);
}
```
This works if, for example, `percentage` is 25 to represent 25%. Some questions in regards to this though:
1. Does PRBMath provide any utility here or should we just do `result = x * percentage / 100`?
2. If we wanted to do a `percentage` lower than 1%, such as 0.1%, this wouldn't be possible with the current method. Is there some way PBRMath could help with getting some more precision?

Thanks in advance!

**Reply by PaulRBerg** [reply]
1. No, PRBMath provides no utility for percentages, you should just do `x * percentage / 1e20` (not 100, because again, PRBMath treats 1e18 as equal to 1, PLEASE read the README).
2. It would be possible to do 0.1%. 0.1% is 1e17. Please read the README.
